### PR TITLE
Read StandardOutput in a smart way to avoid infinite loops

### PR DESCRIPTION
**Pull request by thyams**
Potential fix for SynchronousProcessOutputLoggerWithDiagnoser infinite loop.
Related to my issue in #828 . Detected an abnormal streamread lock/loop. After some investigation, it looks to be potentially an infinite loop at the StreamReader.ReadLine(). It is capable of returning `string.Empty` and `null` as valid values.  The loop proceeds to continue when `string.IsNullOrEmpty`.

If EndOfStream is never achieved/reached, then a break or timeout will need to be added.

Ran the unit test that calls Execute and found that it froze during WaitForExit. I saw the comment should we add a Timeout and proceeded to do so.

Added timeout in Execute, detected freeze during UnitTest.
`ToolchainTests.CustomToolchainsAreSupported()`

* The infinite loop won't be fixed if the stream never ends.
* The infinite loop won't be fixed if the StreamReader can't read the underlying file (OS lock contention).
   * Wanted to have added a 5 second timeout to read the line to assist with this issue, but the underlying stream doesn't support timeouts.

The Execute timeout probably needs to be reviewed and debated about, I have chosen an arbitrary 30 seconds on a WaitForExit().

Commit 2/3, attempted to add timeouts then undo them.

Commit 4 attempted enhancement utilizing StreamReader peek() to see if it can bypass a ReadLine timeout.

Similar Stackoverflow Examples:
https://stackoverflow.com/questions/35318588/streamreader-readline-will-hang-in-an-infinite-loop
https://stackoverflow.com/questions/286533/filestream-streamreader-problem-in-c-sharp
https://stackoverflow.com/questions/13705209/check-if-a-streamreader-can-read-another-line

### CODE REVIEW

File: src/BenchmarkDotNet/Loggers/SynchronousProcessOutputLoggerWithDiagnoser.cs [hunk 1]
```diff
@@ -38,22 +38,30 @@ public SynchronousProcessOutputLoggerWithDiagnoser(ILogger logger, Process proce
...
+                            diagnoser?.Handle(signal, diagnoserActionParameters);
+                            process.StandardInput.WriteLine(Engine.Signals.Acknowledgment);
+                        }
+                        else
+                        { LinesWithExtraOutput.Add(line); }
+                    }
                 }
-                else
-                    LinesWithExtraOutput.Add(line);
+                else { break; } // No more 'characters' can be read.
```
Comment (adamsitnik) [inline src/BenchmarkDotNet/Loggers/SynchronousProcessOutputLoggerWithDiagnoser.cs]: could you please remove the `else { break; }`  and move the condition to while loop?

`while (!process.StandardOutput.EndOfStream && process.StandardOutput.Peek() > 0)`

File: src/BenchmarkDotNet/Toolchains/Executor.cs [hunk 2]
```diff
@@ -69,7 +69,11 @@ private ExecuteResult Execute(Process process, BenchmarkCase benchmarkCase, Sync
 
             loggerWithDiagnoser.ProcessInput();
 
-            process.WaitForExit(); // should we add timeout here?
+            // Timeout Added. Freeze detected.
+            if (!process.WaitForExit(30000))
```
Comment (adamsitnik) [inline src/BenchmarkDotNet/Toolchains/Executor.cs]: 30000 ms is 30 s which is definitely too little for majority of the benchmarks we run.

Please revert this change and create a new issue where we can discuss what the default timeout should be. We should also make it configurable

**Comment by thyams** [comment]
@adamsitnik changes made, off to work for me!

**Comment by adamsitnik** [comment]
@thyams once again big thanks!

From time to time our Travis MacOs build hangs and timeouts. I hope that this change is going to fix this issue too!
